frame/elections-phragmen/src/lib.rs

```diff
@@ -1153,10 +1170,89 @@ impl<T: Config> ContainsLengthBound for Pallet<T> {
 	}
 }
 
+
+impl<T: Config> ElectionDataProvider for Pallet<T> {
+	type AccountId = T::AccountId;
+	type BlockNumber = T::BlockNumber;
+	type MaxVotesPerVoter = ConstU32<MAXIMUM_VOTE>;
+
+	fn electable_targets(
+		maybe_max_len: Option<usize>,
+	) -> frame_election_provider_support::data_provider::Result<Vec<Self::AccountId>> {
+			let mut candidates_and_deposit = Self::candidates();
+			// add all the previous members and runners-up as candidates as well.
+			candidates_and_deposit.append(&mut Self::implicit_candidates_with_deposit());
+			let mut targets: Vec<_> = candidates_and_deposit.into_iter().map(|(target, _)| target).collect();
+
+		Self::register_weight(T::WeightInfo::electable_candidates(targets.len() as u32));
+
+		if let Some(max_candidates) = maybe_max_len {
+			// TODO(gpestana):: or should return Err instead?
+			targets.truncate(max_candidates);
+		}
+
+		Ok(targets)
+	}
+
+	fn electing_voters(
+		maybe_max_len: Option<usize>,
+	) -> frame_election_provider_support::data_provider::Result<
+		Vec<frame_election_provider_support::VoterOf<Self>>,
+	> {
+		let mut voters_and_stakes = Vec::new();
+
+		match Voting::<T>::iter().try_for_each(|(voter, Voter { stake, votes, .. })| {
+			if let Some(max_voters) = maybe_max_len {
+				if voters_and_stakes.len() >= max_voters {
+					return Err(());
+				}
+			}
+			let bounded_votes: BoundedVec<_, Self::MaxVotesPerVoter> =
+				BoundedVec::try_from(votes).map_err(|_| ())?;
+
+			voters_and_stakes.push((voter, stake.unique_saturated_into(), bounded_votes));
+			Ok(())
+		}) {
+			Ok(_) => (),
+			Err(_) => {
+				log::error!(
+					target: "runtime::elections",
+					"Failed to run election. Number of voters exceeded",
+				);
+				Self::deposit_event(Event::ElectionError);
+				Self::register_weight(T::WeightInfo::electing_voters(voters_and_stakes.len() as u32));
+			},
+		}
+
+		Self::register_weight(T::WeightInfo::electing_voters(voters_and_stakes.len() as u32));
```

Comment:
I wonder if we should add the self-weighing logic in `DataProvider::elect()` only and reuse most of the weighting logic of the former `do_phragmen()`. I think that should be OK since I would expect the `DataProvider` to be used only by the election provider. OTOH, if the developers call the `DataProvider` explicitly for some reason, if the weights are implicit under `elect()` we may get into inconsistent weights.